Objects/unicodeobject.c

```diff
@@ -10675,39 +10675,47 @@ PyUnicode_CompareWithASCIIString(PyObject* uni, const char* str)
 
 int
 PyUnicode_EqualToUTF8(PyObject *unicode, const char *str)
+{
+    return PyUnicode_EqualToUTF8AndSize(unicode, str, strlen(str));
+}
+
+int
+PyUnicode_EqualToUTF8AndSize(PyObject *unicode, const char *str, Py_ssize_t size)
 {
     assert(_PyUnicode_CHECK(unicode));
     assert(str);
 
     if (PyUnicode_IS_ASCII(unicode)) {
-        size_t len = (size_t)PyUnicode_GET_LENGTH(unicode);
-        return strlen(str) == len &&
+        Py_ssize_t len = PyUnicode_GET_LENGTH(unicode);
+        return size == len &&
             memcmp(PyUnicode_1BYTE_DATA(unicode), str, len) == 0;
     }
     if (PyUnicode_UTF8(unicode) != NULL) {
-        size_t len = (size_t)PyUnicode_UTF8_LENGTH(unicode);
-        return strlen(str) == len &&
+        Py_ssize_t len = PyUnicode_UTF8_LENGTH(unicode);
+        return size == len &&
             memcmp(PyUnicode_UTF8(unicode), str, len) == 0;
     }
 
-    const unsigned char *s = (const unsigned char *)str;
     Py_ssize_t len = PyUnicode_GET_LENGTH(unicode);
+    if ((size_t)len >= (size_t)size || (size_t)len < (size_t)size / 4) {
+        return 0;
+    }
+    const unsigned char *s = (const unsigned char *)str;
+    const unsigned char *ends = s + (size_t)size;
     int kind = PyUnicode_KIND(unicode);
     const void *data = PyUnicode_DATA(unicode);
     /* Compare Unicode string and UTF-8 string */
     for (Py_ssize_t i = 0; i < len; i++) {
         Py_UCS4 ch = PyUnicode_READ(kind, data, i);
-        if (ch == 0) {
-            return 0;
-        }
-        else if (ch < 0x80) {
-            if (s[0] != ch) {
+        if (ch < 0x80) {
+            if (ends == s || s[0] != ch) {
                 return 0;
             }
             s += 1;
         }
         else if (ch < 0x800) {
-            if (s[0] != (0xc0 | (ch >> 6)) ||
+            if (ends - s < 2 ||
```

Comment:
I prefer to add parenthesis to make the operator priority more obvious:

```suggestion
            if ((ends - s) < 2 ||
```

Same remark for other `ends - s` comparisons below.

Reply:
I prefer to use less parentheses if it is non-ambiguous, because extra parentheses can make the code less readable (does this parenthesis matches this parenthesis or that?). But if you insist...